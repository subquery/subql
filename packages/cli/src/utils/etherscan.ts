// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0

async function runRequest<T>(chainId: string | number, params: Record<string, string>): Promise<T | undefined> {
  const url = `https://api.etherscan.io/v2/api?${new URLSearchParams({
    ...params,
    chainid: String(chainId),
    apiKey: process.env.ETHERSCAN_API_KEY || '',
  }).toString()}`;

  const res = await fetch(url);

  if (!res.ok) {
    throw new Error(`Failed to fetch from Etherscan API: ${res.statusText}`);
  }

  const data = (await res.json()) as {status: '0' | '1'; message: string; result: T};

  if (data.status === '0') {
    if (data.result === 'Missing/Invalid API Key') {
      throw new Error('Missing or invalid Etherscan API key. Please set ETHERSCAN_API_KEY environment variable.');
    }
    if (data.message === 'No data found') {
      return undefined;
    }
    throw new Error(`Etherscan API error: ${data.result}`);
  }

  return data.result;
}

/**
 * Fetches the ABI from Etherscan.
 * @param address - The contract address to fetch the ABI for.
 * @param chainId - The chain ID of the Ethereum network.
 * @returns The ABI as a JSON object or undefined if the fetch fails.
 */
export async function tryFetchAbiFromExplorer(address: string, chainId: number | string): Promise<unknown | undefined> {
  try {
    const result = await runRequest<string>(chainId, {
      module: 'contract',
      action: 'getabi',
      address,
    });

    if (!result) {
      return undefined;
    }

    return JSON.parse(result);
  } catch (e) {
    if (e instanceof Error && e.message.includes('Contract source code not verified')) {
      return undefined;
    }
    throw e;
  }
}

/**
 * Fetches the deployment height of a contract from Etherscan.
 * @param address - The contract address to fetch the deployment height for.
 * @param chainId - The chain ID of the Ethereum network.
 * @returns The block number where the contract was deployed, or undefined if not found.
 */
export async function fetchContractDeployHeight(
  address: string,
  chainId: number | string
): Promise<number | undefined> {
  const result = await runRequest<{blockNumber: string}[]>(chainId, {
    module: 'contract',
    action: 'getcontractcreation',
    contractaddresses: address,
  });

  if (!result) {
    return undefined;
  }

  return parseInt(result[0].blockNumber, 10);
}
