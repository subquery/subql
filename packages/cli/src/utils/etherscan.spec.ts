// Copyright 2020-2025 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0

import {fetchContractDeployHeight, tryFetchAbiFromExplorer} from './etherscan';

describe('etherscan api', () => {
  describe('ABI fetching', () => {
    it('can fetch the ABI', async () => {
      const abi = await tryFetchAbiFromExplorer('0x53A270147832Fc4eb962584911F5D55e86b1BA3F', '8453' /* Base */);
      expect(abi).toMatchSnapshot();
    });

    it('returns undefined if an ABI has not been uploaded', async () => {
      const abi = await tryFetchAbiFromExplorer('0x4eE879f39Cce3C4CA80E2ee90F9Df5aFEEaeb220', '1' /* Ethereum */);
      expect(abi).toBeUndefined();
    });

    it('throws if a contract is non-existant', async () => {
      const abi = await tryFetchAbiFromExplorer('0x53A270147832Fc4eb962584911F5D55e86b1BA3F', '1' /* Ethereum */);
      expect(abi).toBeUndefined();
    });

    it('throws for a network that is not supported', async () => {
      await expect(
        tryFetchAbiFromExplorer('0x53A270147832Fc4eb962584911F5D55e86b1BA3F', '1329' /* Moonbeam */)
      ).rejects.toThrow(
        'Etherscan API error: Missing or unsupported chainid parameter (required for v2 api), please see https://api.etherscan.io/v2/chainlist for the list of supported chainids'
      );
    });
  });

  describe('Deployment information', () => {
    it('can get the deploy height of a contract', async () => {
      const height = await fetchContractDeployHeight('0x53A270147832Fc4eb962584911F5D55e86b1BA3F', '8453' /* Base */);

      expect(height).toBe(22771721);
    });

    it('returns undefined when getting the deploy height for a non-existant contract', async () => {
      const height = await fetchContractDeployHeight('0x53A270147832Fc4eb962584911F5D55e86b1BA3F', '1' /* Ethereum */);
      expect(height).toBeUndefined();
    });
  });
});
