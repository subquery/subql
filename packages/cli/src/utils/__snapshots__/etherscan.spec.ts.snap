// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`etherscan api ABI fetching can fetch the ABI 1`] = `
[
  {
    "inputs": [
      {
        "internalType": "address",
        "name": "tokenAddress",
        "type": "address",
      },
      {
        "internalType": "address",
        "name": "prizePoolAddress",
        "type": "address",
      },
      {
        "internalType": "uint256",
        "name": "poolFeePerc_",
        "type": "uint256",
      },
      {
        "internalType": "address",
        "name": "shieldAddress",
        "type": "address",
      },
      {
        "internalType": "uint256",
        "name": "shieldFeePerc_",
        "type": "uint256",
      },
    ],
    "stateMutability": "nonpayable",
    "type": "constructor",
  },
  {
    "anonymous": false,
    "inputs": [
      {
        "indexed": true,
        "internalType": "address",
        "name": "user",
        "type": "address",
      },
      {
        "indexed": false,
        "internalType": "bytes32",
        "name": "hashedPrompt",
        "type": "bytes32",
      },
      {
        "indexed": false,
        "internalType": "uint256",
        "name": "amount",
        "type": "uint256",
      },
    ],
    "name": "BuyIn",
    "type": "event",
  },
  {
    "stateMutability": "payable",
    "type": "fallback",
  },
  {
    "inputs": [
      {
        "internalType": "bytes32",
        "name": "hashedPrompt",
        "type": "bytes32",
      },
    ],
    "name": "buyIn",
    "outputs": [],
    "stateMutability": "payable",
    "type": "function",
  },
  {
    "inputs": [],
    "name": "operator",
    "outputs": [
      {
        "internalType": "address",
        "name": "",
        "type": "address",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  {
    "inputs": [],
    "name": "poolFeePerc",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  {
    "inputs": [],
    "name": "prizePool",
    "outputs": [
      {
        "internalType": "address",
        "name": "",
        "type": "address",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  {
    "inputs": [
      {
        "internalType": "address",
        "name": "prizePoolAddress",
        "type": "address",
      },
      {
        "internalType": "address",
        "name": "shieldAddress",
        "type": "address",
      },
    ],
    "name": "setAddress",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function",
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "poolFeePerc_",
        "type": "uint256",
      },
      {
        "internalType": "uint256",
        "name": "shieldFeePerc_",
        "type": "uint256",
      },
    ],
    "name": "setFees",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function",
  },
  {
    "inputs": [
      {
        "internalType": "address",
        "name": "operator_",
        "type": "address",
      },
    ],
    "name": "setOperator",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function",
  },
  {
    "inputs": [],
    "name": "shield",
    "outputs": [
      {
        "internalType": "address",
        "name": "",
        "type": "address",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  {
    "inputs": [],
    "name": "shieldFeePerc",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  {
    "inputs": [],
    "name": "token",
    "outputs": [
      {
        "internalType": "contract Token",
        "name": "",
        "type": "address",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
  {
    "stateMutability": "payable",
    "type": "receive",
  },
]
`;
